File: test/ModularPipelines.UnitTests/ExecutionHintTests.cs
using System.Collections.Concurrent;
using ModularPipelines.Attributes;
using ModularPipelines.Context;
using ModularPipelines.Enums;
using ModularPipelines.Modules;
using ModularPipelines.TestHelpers;

namespace ModularPipelines.UnitTests;

[TUnit.Core.NotInParallel(nameof(ExecutionHintTests))]
public class ExecutionHintTests : TestBase
{
    private static readonly ConcurrentBag<string> CpuModulesExecuting = new();
    private static readonly ConcurrentBag<string> CpuViolations = new();
    private static int _maxCpuConcurrency = 0;

    [ExecutionHint(ExecutionType.CpuIntensive)]
    public class CpuIntensiveModule1 : Module<string>
    {
        public override async Task<string?> ExecuteAsync(IModuleContext context, CancellationToken cancellationToken)
        {
            var moduleName = GetType().Name;
            CpuModulesExecuting.Add(moduleName);

            var currentCount = CpuModulesExecuting.Count;
            if (currentCount > _maxCpuConcurrency)
            {
                Interlocked.Exchange(ref _maxCpuConcurrency, currentCount);
            }

            await Task.Delay(50, cancellationToken);

            // Record violation if more than allowed
            if (CpuModulesExecuting.Count > 2)
            {
                CpuViolations.Add($"{moduleName}: {CpuModulesExecuting.Count} concurrent CPU-intensive modules");
            }

            CpuModulesExecuting.TryTake(out _);
            return moduleName;
        }
    }

    [ExecutionHint(ExecutionType.CpuIntensive)]
    public class CpuIntensiveModule2 : Module<string>
    {
        public override async Task<string?> ExecuteAsync(IModuleContext context, CancellationToken cancellationToken)
        {
            var moduleName = GetType().Name;
            CpuModulesExecuting.Add(moduleName);

            var currentCount = CpuModulesExecuting.Count;
            if (currentCount > _maxCpuConcurrency)
            {
                Interlocked.Exchange(ref _maxCpuConcurrency, currentCount);
            }

            await Task.Delay(50, cancellationToken);

            CpuModulesExecuting.TryTake(out _);
            return moduleName;
        }
    }

    [ExecutionHint(ExecutionType.CpuIntensive)]
    public class CpuIntensiveModule3 : Module<string>
    {
        public override async Task<string?> ExecuteAsync(IModuleContext context, CancellationToken cancellationToken)
        {
            var moduleName = GetType().Name;
            CpuModulesExecuting.Add(moduleName);

            var currentCount = CpuModulesExecuting.Count;
            if (currentCount > _maxCpuConcurrency)
            {
                Interlocked.Exchange(ref _maxCpuConcurrency, currentCount);
            }

            await Task.Delay(50, cancellationToken);

            CpuModulesExecuting.TryTake(out _);
            return moduleName;
        }
    }

    [ExecutionHint(ExecutionType.IoIntensive)]
    public class IoIntensiveModule : Module<string>
    {
        public override async Task<string?> ExecuteAsync(IModuleContext context, CancellationToken cancellationToken)
        {
            await Task.Delay(10, cancellationToken);
            return "IoIntensive";
        }
    }

    [ExecutionHint(ExecutionType.Default)]
    public class DefaultExecutionTypeModule : Module<string>
    {
        public override Task<string?> ExecuteAsync(IModuleContext context, CancellationToken cancellationToken)
        {
            return Task.FromResult<string?>("Default");
        }
    }

    public class NoHintModule : Module<string>
    {
        public override Task<string?> ExecuteAsync(IModuleContext context, CancellationToken cancellationToken)
        {
            return Task.FromResult<string?>("NoHint");
        }
    }

    [Before(Test)]
    public void ClearState()
    {
        while (CpuModulesExecuting.TryTake(out _)) { }
        while (CpuViolations.TryTake(out _)) { }
        _maxCpuConcurrency = 0;
    }

    [Test]
    public async Task ExecutionHintAttribute_CanBeAppliedToModule()
    {
        var result = await TestPipelineHostBuilder.Create()
            .AddModule<CpuIntensiveModule1>()
            .ExecutePipelineAsync();

        await Assert.That(result.Status).IsEqualTo(Status.Successful);
    }

    [Test]
    public async Task ModulesWithoutExecutionHint_UseDefaultType()
    {
        var result = await TestPipelineHostBuilder.Create()
            .AddModule<NoHintModule>()
            .ExecutePipelineAsync();

        await Assert.That(result.Status).IsEqualTo(Status.Successful);
    }

    [Test]
    public async Task AllExecutionTypes_ExecuteSuccessfully()
    {
        var result = await TestPipelineHostBuilder.Create()
            .AddModule<CpuIntensiveModule1>()
            .AddModule<IoIntensiveModule>()
            .AddModule<DefaultExecutionTypeModule>()
            .AddModule<NoHintModule>()
            .ExecutePipelineAsync();

        await Assert.That(result.Status).IsEqualTo(Status.Successful);
    }

    [Test]
    public async Task CpuIntensiveModules_AreThrottled()
    {
        // Set max CPU-intensive modules to 2
        var result = await TestPipelineHostBuilder.Create()
            .AddModule<CpuIntensiveModule1>()
            .AddModule<CpuIntensiveModule2>()
            .AddModule<CpuIntensiveModule3>()
            .ConfigurePipelineOptions((_, options) =>
            {
                options.Concurrency.MaxCpuIntensiveModules = 2;
            })
            .ExecutePipelineAsync();

        await Assert.That(result.Status).IsEqualTo(Status.Successful);
        // The max concurrency should not exceed 2
        await Assert.That(_maxCpuConcurrency).IsLessThanOrEqualTo(2);
    }
}

File: test/ModularPipelines.UnitTests.FSharp/Execution/ExecutionHintTests.fs
namespace ModularPipelines.UnitTests.FSharp.Execution

open System.Collections.Concurrent
open System.Threading
open System.Threading.Tasks
open ModularPipelines.Attributes
open ModularPipelines.Context
open ModularPipelines.Enums
open ModularPipelines.Extensions
open ModularPipelines.Modules
open ModularPipelines.TestHelpers
open TUnit.Assertions
open TUnit.Assertions.FSharp.Operations
open TUnit.Core
open TUnit.Assertions.Extensions

module private SharedState =
    let cpuModulesExecuting = ConcurrentBag<string>()
    let cpuViolations = ConcurrentBag<string>()
    let mutable maxCpuConcurrency = 0

[<ExecutionHint(ExecutionType.CpuIntensive)>]
type private CpuIntensiveModule1() =
    inherit Module<string>()

    override this.ExecuteAsync(_: IModuleContext, cancellationToken: CancellationToken) =
        task {
            let moduleName = this.GetType().Name
            SharedState.cpuModulesExecuting.Add(moduleName.ToString())

            let currentCount = SharedState.cpuModulesExecuting.Count
            if currentCount > SharedState.maxCpuConcurrency then
                Interlocked.Exchange(&SharedState.maxCpuConcurrency, currentCount) |> ignore

            do! Task.Delay(50, cancellationToken)

            if SharedState.cpuModulesExecuting.Count > 2 then
                SharedState.cpuViolations.Add($"{moduleName}: {SharedState.cpuModulesExecuting.Count} concurrent CPU-intensive modules")

            let mutable ignored = Unchecked.defaultof<string>
            SharedState.cpuModulesExecuting.TryTake(&ignored) |> ignore
            return moduleName
        }

[<ExecutionHint(ExecutionType.CpuIntensive)>]
type private CpuIntensiveModule2() =
    inherit Module<string>()

    override this.ExecuteAsync(_: IModuleContext, cancellationToken: CancellationToken) =
        task {
            let moduleName = this.GetType().Name
            SharedState.cpuModulesExecuting.Add(moduleName)

            let currentCount = SharedState.cpuModulesExecuting.Count
            if currentCount > SharedState.maxCpuConcurrency then
                Interlocked.Exchange(&SharedState.maxCpuConcurrency, currentCount) |> ignore

            do! Task.Delay(50, cancellationToken)

            let mutable ignored = Unchecked.defaultof<string>
            SharedState.cpuModulesExecuting.TryTake(&ignored) |> ignore
            return moduleName
        }

[<ExecutionHint(ExecutionType.CpuIntensive)>]
type private CpuIntensiveModule3() =
    inherit Module<string>()

    override this.ExecuteAsync(_: IModuleContext, cancellationToken: CancellationToken) =
        task {
            let moduleName = this.GetType().Name
            SharedState.cpuModulesExecuting.Add(moduleName)

            let currentCount = SharedState.cpuModulesExecuting.Count
            if currentCount > SharedState.maxCpuConcurrency then
                Interlocked.Exchange(&SharedState.maxCpuConcurrency, currentCount) |> ignore

            do! Task.Delay(50, cancellationToken)

            let mutable ignored = Unchecked.defaultof<string>
            SharedState.cpuModulesExecuting.TryTake(&ignored) |> ignore
            return moduleName
        }

[<ExecutionHint(ExecutionType.IoIntensive)>]
type private IoIntensiveModule() =
    inherit Module<string>()
    override _.ExecuteAsync(_: IModuleContext, cancellationToken: CancellationToken) =
        task {
            do! Task.Delay(10, cancellationToken)
            return "IoIntensive"
        }

[<ExecutionHint(ExecutionType.Default)>]
type private DefaultExecutionTypeModule() =
    inherit Module<string>()
    override _.ExecuteAsync(_: IModuleContext, _: CancellationToken) = Task.FromResult("Default")

type private NoHintModule() =
    inherit Module<string>()
    override _.ExecuteAsync(_: IModuleContext, _: CancellationToken) = Task.FromResult("NoHint")

[<NotInParallel(nameof ExecutionHintTests)>]
type ExecutionHintTests() =
    inherit TestBase()

    [<Before(HookType.Test)>]
    member _.ClearState() =
        let mutable executingItem = Unchecked.defaultof<string>
        while SharedState.cpuModulesExecuting.TryTake(&executingItem) do ()

        let mutable violationItem = Unchecked.defaultof<string>
        while SharedState.cpuViolations.TryTake(&violationItem) do ()

        SharedState.maxCpuConcurrency <- 0

    [<Test>]
    member _.ExecutionHintAttribute_CanBeAppliedToModule() = async {
        let! result =
            TestPipelineHostBuilder.Create().AddModule<CpuIntensiveModule1>().ExecutePipelineAsync()
            |> Async.AwaitTask

        do! check(Assert.That(result.Status).IsEqualTo(Status.Successful))
    }

    [<Test>]
    member _.ModulesWithoutExecutionHint_UseDefaultType() = async {
        let! result =
            TestPipelineHostBuilder.Create().AddModule<NoHintModule>().ExecutePipelineAsync()
            |> Async.AwaitTask

        do! check(Assert.That(result.Status).IsEqualTo(Status.Successful))
    }

    [<Test>]
    member _.AllExecutionTypes_ExecuteSuccessfully() = async {
        let! result =
            TestPipelineHostBuilder.Create()
                .AddModule<CpuIntensiveModule1>()
                .AddModule<IoIntensiveModule>()
                .AddModule<DefaultExecutionTypeModule>()
                .AddModule<NoHintModule>()
                .ExecutePipelineAsync()
            |> Async.AwaitTask

        do! check(Assert.That(result.Status).IsEqualTo(Status.Successful))
    }

    [<Test>]
    member _.CpuIntensiveModules_AreThrottled() = async {
        let! result =
            TestPipelineHostBuilder.Create()
                .AddModule<CpuIntensiveModule1>()
                .AddModule<CpuIntensiveModule2>()
                .AddModule<CpuIntensiveModule3>()
                .ConfigurePipelineOptions(fun _ options -> options.Concurrency.MaxCpuIntensiveModules <- 2)
                .ExecutePipelineAsync()
            |> Async.AwaitTask

        do! check(Assert.That(result.Status).IsEqualTo(Status.Successful))
        do! check(Assert.That(SharedState.maxCpuConcurrency).IsLessThanOrEqualTo(2))
    }
